modules/artifact/src/main/java/org/mule/module/artifact/classloader/GroovyResourceReleaser.java

```diff
@@ -0,0 +1,116 @@
+  private static final String GROOVY_CLASS_INFO = "org.codehaus.groovy.reflection.ClassInfo";
+  private static final String GROOVY_INVOKER_HELPER = "org.codehaus.groovy.runtime.InvokerHelper";
+  private static final String GROOVY_SCRIPT_ENGINE_FACTORY = "org.codehaus.groovy.jsr223.GroovyScriptEngineFactory";
+  private static final String LOGGER_ABSTRACT_MANAGER = "org.apache.logging.log4j.core.appender.AbstractManager";
+  private static final String LOGGER_ROLLING_FILE_MANAGER = "org.apache.logging.log4j.core.appender.rolling.RollingFileManager";
+  private static final String LOGGER_CONFIGURATION = "org.apache.logging.log4j.core.config.Configuration";
+
+  /**
+   * Creates a new Instance of GroovyResourceReleaser
+   *
+   * @param classLoader ClassLoader which loaded the Groovy Dependency.
+   */
+  public GroovyResourceReleaser(ClassLoader classLoader) {
+    this.classLoader = classLoader;
+  }
+
+  @Override
+  public void release() {
+    unregisterInvokerHelper();
+    cleanSpisEngines();
+  }
+
+  private void unregisterInvokerHelper() {
+    try {
+      Class classInfoClass = this.classLoader.loadClass(GROOVY_CLASS_INFO);
+      Method getAllClassInfoMethod = classInfoClass.getMethod("getAllClassInfo");
+      Method getTheClassMethod = classInfoClass.getMethod("getTheClass");
+      Class invokerHelperClass = this.classLoader.loadClass(GROOVY_INVOKER_HELPER);
+      Method removeClassMethod = invokerHelperClass.getMethod("removeClass", Class.class);
+      Object classes = getAllClassInfoMethod.invoke(null);
+      if (classes != null && classes instanceof Collection) {
+        for (Object classInfo : ((Collection) classes)) {
+          String className = "";
+          try {
+            Object clazz = getTheClassMethod.invoke(classInfo);
+            Method getNameMethod = clazz.getClass().getMethod("getName");
+            className = (String) getNameMethod.invoke(clazz);
+            removeClassMethod.invoke(null, clazz);
+          } catch (IllegalAccessException | InvocationTargetException e) {
+            LOGGER.warn("Could not remove the {} class from the Groovy's InvokerHelper", className, e);
+          }
+        }
+      }
+    } catch (IllegalAccessException | InvocationTargetException | ClassNotFoundException | NoSuchMethodException e) {
+      LOGGER.warn("Error trying to remove the Groovy's InvokerHelper classes", e);
+    }
+  }
+
+  private void cleanSpisEngines() {
+    try {
+      Class<?> abstractManager = loadClass(LOGGER_ABSTRACT_MANAGER, this.classLoader);
+      HashMap hashMap = getStaticFieldValue(abstractManager, "MAP", true);
+      Iterator it = hashMap.values().iterator();
+      Class<?> rollingFileManagerClass = loadClass(LOGGER_ROLLING_FILE_MANAGER, this.classLoader);
+      Object rfmInstance = null;
+      while (it.hasNext()) {
+        Object o = it.next();
+        if (rollingFileManagerClass.isInstance(o)) {
```

Comment:
It concerns me a little bit that we are only doing this for the `RollingFileManager`. Couldn't we be overfitting to this particular customer case here? Have we considered other sources of this same leak?
I mean, what if the customer is using some other kind of layout which also uses scripting.
I haven't looked at Log4j's  code but it seems we are just trying to get our hands on the `Configuration` regardless of the layout.

Similarly, how can we assume that this leak will only happen with groovy and not with other scripting engines? I mean, Java's ScriptEngineManager, through Log4j's `ScriptManager`, will probably have retained any scripting engine that is used, right?
Is it because only `GroovyScriptEngineFactory` will indirectly retain the artifact's `ClassLoader`? If that is truly the case, we should at least leave some comment here.

Reply:
Sorry, I forgot to mention something else. Are we sure we need to use reflection to manipulate the Log4j classes? I get it that we may need it for accessing the private stuff, but the classes and public methods should be available to the ClassLoader that loaded this releaser Class, right? Maybe I'm missing something, but it sounds worth checking because it would make everything clearer and less error prone.

Reply:
It's true. RollingFileManager is just an implementation. I've changed it to be more generic.